bundler/src/common.ts

```diff
@@ -14,4 +14,214 @@
-export const BUNDLE_ANNOTATION_REGEX = /bundles.validator.forsetisecurity.org\/(.+)/
+import {
+  existsSync,
+  mkdirSync,
+  readFileSync,
+  unlinkSync,
+  writeFileSync
+} from "fs";
+import * as glob from "glob";
+import { DumpOptions, safeDump } from 'js-yaml';
+import { KubernetesObject, getAnnotation } from "kpt-functions";
+import * as path from "path";
+
+export const BUNDLE_ANNOTATION_REGEX = /bundles.validator.forsetisecurity.org\/(.+)/;
+export const CT_KIND = "ConstraintTemplate";
+export const SUPPORTED_API_VERSIONS = /^(constraints|templates).gatekeeper.sh\/v1(.+)$/;
+const YAML_STYLE: DumpOptions = {
+  // indentation width to use (in spaces).
+  indent: 2,
+  // when true, will not add an indentation level to array elements.
+  noArrayIndent: true,
+};
+
+class PolicyLibrary {
+  bundles: Map<string, PolicyBundle>;
+  configs: KubernetesObject[];
+
+  constructor(configs: KubernetesObject[]) {
+    this.bundles = new Map();
+    this.configs = new Array();
+
+    configs
+      .filter(o => {
+        return PolicyConfig.isPolicyObject(o);
+      })
+      .forEach(o => {
+        const annotations = o.metadata.annotations || {};
+        Object.keys(annotations).forEach(annotation => {
+          const result = annotation.match(BUNDLE_ANNOTATION_REGEX);
+          if (!result) {
+            return;
+          }
+          const bundle = result[0];
+          const control = annotations[annotation];
+          this.bundlePolicy(bundle, control, o);
+        });
+        this.configs.push(o);
+      });
+  }
+
+  getAll(): KubernetesObject[] {
+    return this.configs;
+  }
+
+  getTemplates(): KubernetesObject[] {
+    return this.getOfKind(CT_KIND);
+  }
+
+  getTemplate(kind: string): KubernetesObject {
+    const matches = this.getTemplates().filter(o => {
+      return (o as any).spec.crd.spec.names.kind === kind;
+    });
+    return matches[0] || undefined;
+  }
+
+  getOfKind(kind: string): KubernetesObject[] {
+    return this.configs.filter(o => {
+      return o.kind === kind;
+    });
+  }
+
+  bundlePolicy(bundleKey: string, control: string, policy: KubernetesObject) {
+    let bundle = this.bundles.get(bundleKey);
+    if (bundle === undefined) {
+      bundle = new PolicyBundle(bundleKey);
+      this.bundles.set(bundleKey, bundle);
+    }
+
+    bundle.addPolicy(policy);
+  }
+}
+
+class PolicyBundle {
+  key: string;
+  configs: KubernetesObject[];
+
+  constructor(annotation: string) {
+    this.key = annotation;
+    this.configs = new Array();
+  }
+
+  getName() {
+    const matches = this.key.match(BUNDLE_ANNOTATION_REGEX);
+    return matches ? matches[1] : "Unknown";
+  }
+
+  getKey() {
+    return this.getName();
+  }
+
+  addPolicy(policy: KubernetesObject) {
+    this.configs.push(policy);
+  }
+
+  getConfigs() {
+    return this.configs;
+  }
+
+  getControl(policy: KubernetesObject): string {
+    return getAnnotation(policy, this.key) || "";
+  }
+
+  write(sinkDir: string, overwrite: boolean) {
```

Comment:
The reason I wanted to add that in is because otherwise if the sink dir is not-empty, then they wouldn't get a "clean" bundle(s). If you don't think we should provide the feature, then I can remove it and have people use kpt sink.

Reply:
I think for now let's not overwrite existing libraries.

Reply:
Updated. If anyone wants that ability they can actually just pipe the output to the [write-yaml demo function](https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/demo-functions/src/write_yaml.ts) as well.